### PR TITLE
Game parser change

### PR DESCRIPTION
**Pull request by RoiEXLab**
This is some sort of test, with "unimportant parsing".
Basically I replaced the DOM parser with a SAX parser in the map creator code.
The classes are supposed to be used from the current GameParser in another PR, I wasn't sure if this would work out as expected.
Testxml: war_of_the_relics
Time in miliseconds:
```
//new
466
464
467
445
475

//old
588
581
573
561
```
Of course the "improvement" will differ when applying this to the in-game parser

Changes are easier reviewed with whitespace ignored `?w=1`

### CODE REVIEW

File: src/main/java/games/strategy/engine/data/TripleaHandler.java [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+import com.google.common.collect.Multimap;
+
+public abstract class TripleaHandler extends DefaultHandler {
+
+  private final GameData gameData;
+  private final Deque<PseudoElement> stack = new LinkedList<>();
+  private static final Multimap<String, String> requiredParents = HashMultimap.create();
+
+  static {
+    requiredParents.put(XML_NODE_NAME_GAME, null);
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: This map checks the validity of the xml file

**Comment by RoiEXLab** [comment]
Oh and btw, I tested the parser with a couple of maps, but not all of them. Please tell me if you run into any issues.

**Comment by ssoloff** [comment]
@RoiEXLab Just for clarification...  Is this PR intended to be merged or are you just soliciting feedback for another PR (that's basically how I read your description above).  If the latter, I don't want to waste your time asking questions like, "why are you referencing code from `tools.map.xml.creator.*` in the engine?" :smile:

**Comment by RoiEXLab** [comment]
@ssoloff Any idea why the travis build is failing because of checkstyle, because codacy tells me everything is fine?

**Comment by RoiEXLab** [comment]
@ssoloff Both, if this works out I will submit another PR which changes the actual game code.
But this PR can already be tested now, so mistakes in the follow-up PR are going to be less likely

**Comment by ssoloff** [comment]
@RoiEXLab

> Any idea why the travis build is failing because of checkstyle, because codacy tells me everything is fine?

It looks like your problem is that Travis is building your branch merged with `master`, while your local build (and presumably Codacy) are just looking at the PR branch.

As of #2035, we now require Javadocs on public types.  That requirement is probably being pulled in when Travis checks out the PR code merged with `master`.  You have a couple of new types in this PR, so they are most likely responsible for causing the warning count to exceed the threshold.

I would suggest rebasing with the latest `master` to get the new Checkstyle configuration.  (Note that you will have to restart Eclipse after rebasing for the Checkstyle plugin to pick up the new configuration.)

File: src/main/java/games/strategy/engine/data/GameParseException.java [hunk 2]
```diff
@@ -1,6 +1,8 @@
 package games.strategy.engine.data;
 
-public class GameParseException extends Exception {
+import org.xml.sax.SAXException;
+
+public class GameParseException extends SAXException {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/GameParseException.java]: Inheriting from `SAXException` is a leaky abstraction.  Clients of `GameParser` shouldn't know or care that we're using a SAX, DOM, XPP, etc. parser.  `SAXException`s thrown internally to the parser should be wrapped in a `GameParseException` and rethrown.

File: src/main/java/games/strategy/engine/data/PseudoElement.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package games.strategy.engine.data;
+
+import org.xml.sax.Attributes;
+
+import com.google.common.base.Joiner;
+
+public class PseudoElement {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/PseudoElement.java]: This class is only used by `TripleaHandler`.  At a minimum, it should be package-private, but it could work also as a private nested class of `TripleaHandler`.

File: src/main/java/games/strategy/engine/data/TripleaHandler.java [hunk 4]
```diff
@@ -0,0 +1,393 @@
+package games.strategy.engine.data;
+
+import static tools.map.xml.creator.MapXmlHelper.XML_NODE_NAME_ALLIANCE;
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: There probably should be no dependency from `games.strategy.*` to `tools.*`.  Vice versa seems to be acceptable, as that dependency path is already present in multiple locations within the tool code.

File: src/main/java/games/strategy/engine/data/TripleaHandler.java [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+import org.xml.sax.Attributes;
+import org.xml.sax.SAXException;
+import org.xml.sax.SAXParseException;
+import org.xml.sax.helpers.AttributesImpl;
+import org.xml.sax.helpers.DefaultHandler;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+
+public abstract class TripleaHandler extends DefaultHandler {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: The only implementation of this class is in `tools.map.creator.xml.MapXmlHelper`.  Is the intention that there will be multiple handlers for game XML files?  Don't they all have the same parsing and validation requirements?

File: src/main/java/tools/map/xml/creator/MapXmlHelper.java [hunk 6]
```diff
@@ -456,438 +463,327 @@ static void clearGameSettings() {
...
+    Wrapper<GameStep> stepToGo = new Wrapper<>(MapXmlCreator.GAME_STEP_FIRST);
 
-    final Node mainlastChild = dom.getLastChild();
-    if (!mainlastChild.getNodeName().equals(XML_NODE_NAME_GAME)) {
-      throw new IllegalArgumentException(
-          "Last node of XML document is not the expeced 'game' node, but '" + mainlastChild.getNodeName() + "'");
-    }
-    return parseGameNode(mainlastChild);
-  }
+    new GameParser(gameXmlPath).parse(new FileInputStream(gameXmlPath), new TripleaHandler(null) {
```
Comment (ssoloff) [inline src/main/java/tools/map/xml/creator/MapXmlHelper.java]: This `TripleaHandler` implementation currently has 0% test coverage.  I'm assuming that's because you haven't replaced calls to `GameParser#parse(InputStream, AtomicReference<String>, boolean)` in the unit tests with `GameParser#parse(InputStream, DefaultHandler)` yet.  It would be interesting to see how much test coverage that will provide.

Since we're introducing a totally new parser, we probably want to ensure high instruction and branch coverage.  `GameParser` currently has 66% instruction and 55% branch coverage, so that should be the minimum we strive for.

File: src/main/java/games/strategy/engine/data/TripleaHandler.java (hunk 4 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: Agreed.
I will move the constants to a separate class

File: src/main/java/games/strategy/engine/data/TripleaHandler.java (hunk 5 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: This is the intention.
Because they are not exactly the same.
MapXmlHelper wants to parse the xml into a GUI which edits those values. Tge game parser wants to translate it into a gamedata object which can be used to run the actual game.
Everything which is the same is supposed to be handled inside the abstract Tripleahandler class

File: src/main/java/games/strategy/engine/data/PseudoElement.java (hunk 3 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/PseudoElement.java]: This class is also used by MapXmlHelper

Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/PseudoElement.java]: Ah, I missed that it was the return type of `TripleaHandler#getCurrentParent()`, which is protected.  In that case, should `PseudoElement` be a protected nested class within `TripleaHandler` since it's only intended to be used by subclasses of `TripleaHandler`?

File: src/main/java/games/strategy/engine/data/TripleaHandler.java [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+          break;
+        case XML_NODE_NAME_PROPERTY_LIST:
+          handlePropertyList();
+          break;
+        default:
+          break;
+      }
+    } catch (GameParseException e) {
+      throw new SAXException(e);
+    }
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: Now I understand why you were inclined to have `GameParseException` extend `SAXException`.  However, I think it's ok for all of the `TripleaHandler` template methods to throw `SAXException` instead of `GameParseException`.

`GameParseException` seems to be used by `GameParser` to communicate domain-specific errors not necessarily related to parsing (e.g. game data validation).  Therefore, it should probably only be thrown by the public methods of `GameParser`.

On the other hand, `TripleaHandler` is effectively part of the SAX parser.  Therefore, its domain-specific exception type is naturally `SAXException`.  Simply changing all occurrences of `GameParseException` to `SAXException` in `TripleaHandler` (and its anonymous subclass in `MapXmlHelper`) seems to work and allows you to get rid of the recently-added try-catch blocks that wrap a `GameParseException` in a `SAXException`.

One could argue that `GameParser#parse()` shouldn't even throw `SAXException`, and that it should always be wrapped in a `GameParseException`, but that's probably outside the scope of this PR.

File: src/main/java/games/strategy/engine/data/TripleaHandler.java [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+        throw new GameParseException(
+            "Parent of '" + qname + "' must be one of " + Arrays.toString(parents.toArray()) + ", but was '"
+                + actualName + "'");
+      }
+    } else if (!qname.equals(XML_NODE_NAME_GAME)) {
+      throw new GameParseException("Root element must be 'game', found '" + qname + "' instead");
+    }
+  }
+
+  protected abstract void handleInfo(Attributes attributes) throws GameParseException;
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: Do all of these `handle*()` methods need to be abstract?  The one implementation in `MapXmlHelper` seems to have a lot of empty methods (20%).  That suggests handling many aspects of the game XML file might be optional.

File: src/main/java/tools/map/xml/creator/MapXmlHelper.java [hunk 9]
```diff
@@ -456,438 +462,328 @@ static void clearGameSettings() {
...
+
+      private CanalTerritoriesTuple canalDef = null;
+      private String newCanalName = null;
+      private SortedSet<String> newLandTerritories = new TreeSet<>();
+      private String attachmentAttachTo = null;
+
+      private List<String> settingValues = new ArrayList<>();
+
+      @Override
+      protected void handleInfo(Attributes attributes) throws GameParseException {
```
Comment (ssoloff) [inline src/main/java/tools/map/xml/creator/MapXmlHelper.java]: Note that none of the `handle*()` methods in this implementation throw any checked exceptions (`GameParseException` or `SAXException`), so the exception specifications can be removed from these methods to reduce noise.

File: src/main/java/games/strategy/engine/data/TripleaHandler.java (hunk 7 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: I'd argue that SAXParsingexceptions should only occur when the xml syntax is malformed, GameParseExceptions include invalid xml content

File: src/main/java/games/strategy/engine/data/TripleaHandler.java (hunk 8 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: They don't, but I'm assuming the real game parser uses almost all of those methods.

File: src/main/java/tools/map/xml/creator/MapXmlHelper.java (hunk 9 above)
Comment (ssoloff) [inline src/main/java/tools/map/xml/creator/MapXmlHelper.java]: Sorry if I wasn't clear about this...  I'm not suggesting to remove the exception specification from the **base class** methods for the reasons you specified.  I'm simply saying that the **subclass** methods do not throw a `GameParseException`, and thus they don't need to include the exception specification.  Since the subclass is anonymous, no other class can override these methods, and the "generic" exception specifications don't need to be present.

Not a big deal.  It's just convenient to be able to look at the signature of a concrete method and tell at a glance what checked exceptions it throws.  In this case, one has to examine the code before coming to the conclusion that, in fact, no checked exceptions are thrown.

Comment (RoiEXLab) [inline src/main/java/tools/map/xml/creator/MapXmlHelper.java]: Ah ok, I will change that

File: src/main/java/games/strategy/engine/data/PseudoElement.java (hunk 3 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/PseudoElement.java]: Good Point

**Comment by ssoloff** [comment]
I'm ok with the latest changes.  I believe the issue about new/existing exception types will become clearer in follow-up PRs.  But since this is an infrastructure change, I think either @ron-murhammer or @DanVanAtta need to review before merging.

File: src/main/java/games/strategy/engine/data/TripleaHandler.java [hunk 10]
```diff
@@ -0,0 +1,475 @@
...
+import org.xml.sax.helpers.DefaultHandler;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+
+/**
+ * Abstract common implementation of a DefaultHandler.
+ */
+public abstract class TripleaHandler extends DefaultHandler {
```
Comment (DanVanAtta) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: - Naming nit: this should be `TripleAHandler`, no?
- `Handler` gives me the heeby jeebies. Reminds me of the kingdom of nouns, and does not really describe anything. perhaps `TripleASaxParser` would be a better name for this class.

File: src/main/java/games/strategy/engine/data/TripleaHandler.java [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+
+  public static final String XML_NODE_NAME_ALLIANCE = "alliance";
```
Comment (DanVanAtta) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: Using an enum instead of String I think would help simplify quite a bit. I'm not sure if you'll be able to capture the parent-child relationship more cleanly, but there will be at least a strong type for node names, we won't need the redundant `XML_NODE_NAME` prefix anymore, and we'll have a strong value type instead of string and can still capture the `label -> string` relationship.

File: src/main/java/games/strategy/engine/data/TripleaHandler.java [hunk 12]
```diff
@@ -0,0 +1,475 @@
...
+      }
+    } catch (GameParseException e) {
+      throw new SAXException(e);
+    }
+  }
+
+  private void checkParentNode(String qname) throws GameParseException {
+    if (!stack.isEmpty()) {
+      String actualName = stack.peek().getName();
+      Collection<String> parents = requiredParents.get(qname);
```
Comment (DanVanAtta) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: There is a good bit of code to define the parent/child relationship, and I wonder about the overhead to understand/modify and maintain those relationships. If we did not check tag nesting structure here, and did not have nor maintain the `requiredParents` data, what would be the impact?

File: src/main/java/games/strategy/engine/data/TripleaHandler.java [hunk 13]
```diff
@@ -0,0 +1,475 @@
...
+                + actualName + "'");
+      }
+    } else if (!qname.equals(XML_NODE_NAME_GAME)) {
+      throw new GameParseException("Root element must be 'game', found '" + qname + "' instead");
+    }
+  }
+
+  protected abstract void handleInfo(Attributes attributes) throws GameParseException;
+
+  protected abstract void handleLoader(Attributes attributes) throws GameParseException;
```
Comment (DanVanAtta) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: I think we can gain a bit by applying open-close here, and could remove the inheritance heirarchy that these abstract methods would require. Noting that there is no `TripleaHandlerTest.java`, there is evidence that this was difficult to test and open-close should help fix that here.

Noting we have a common param of `Attributes`, and above we are mapping strings to these methods, it seems then that we could create a class like:
```
public interface XmlAttributeParser {
  String getXmlNodeName(); // preferably this would return an enum, but as an example in context it would be a String.
  void handle(Attributes attributes) throws GameParseException;
}
```

We could then update the constructor to init a list of these:
```
public TripleAHandler(GameData gameData) {
  this(
     gameData,
     new DiceSidesParser(),
     new TerritoryParser(),
     new connectionParser(),
     new ResourceParser(), 
    :
    :
}
@VisibleForTesting
TripleAHandler(GameData gameData, XmlAttributeParser ... parsers) {
   this.gameData = gameData;
   this.parsers = Arrays.asList(parsers);
}
```

Then handleElements would be simpler and not need to update when we change parsers:
```
 private void handleElement(String qname, Attributes attributes) throws SAXException {
    parsers.stream().
        .filter(parser -> parser.getXmlNodeName.equals(qname))
        .findFirst(parser -> parser.handle(attributes)
        .orElseThrow(() -> new RuntimeException("could not find parser for: " + qname + ", parsers loaded: " + parsers));
 }
```

Finally, the test burden for each parser would be much less, we could test each of the parsers independently and would only need to test the framework/handler code once. This would let us more thoroughly test the error/edge cases of each component as well. With the framework and handlers coupled, each handler needs to be tested via every path that the framework can take, meaning any handler test cases are multiplied by each such path. Splitting those two components allows us to test each individually and their respective test cases would no longer be related.

File: src/main/java/tools/map/xml/creator/MapXmlHelper.java [hunk 14]
```diff
@@ -456,438 +461,328 @@ static void clearGameSettings() {
...
-                    .getNodeValue().substring(3));
-          }
-        }
-        getProductionFrontiersMap().put(playerName, frontierRules);
+      @Override
+      protected void handlePlayer(Attributes attributes) {
+        final String playerNameAttr = attributes.getValue(XML_ATTR_PLAYER_NAME_NAME);
+        getPlayerNames().add(playerNameAttr);
+        getPlayerInitResourcesMap().put(playerNameAttr, 0);
+        // TODO: add logic for optional value
```
Comment (DanVanAtta) [inline src/main/java/tools/map/xml/creator/MapXmlHelper.java]: Is this TODO still outstanding?  Are we ever really going to get to it, or when would we need it? I suspect we could remove this line and the commented out code below.

File: src/main/java/tools/map/xml/creator/MapXmlHelper.java [hunk 15]
```diff
@@ -456,438 +461,328 @@ static void clearGameSettings() {
...
+                Strings.nullToEmpty(attributes.getValue(XML_ATTR_STEP_NAME_PLAYER)),
+                (maxRunCount == null ? 0 : Integer.parseInt(maxRunCount))));
+      }
+
+      @Override
+      protected void handlePlayerProduction(Attributes attributes) {}
+
+      @Override
+      protected void handleCost(Attributes attributes) {
+        costQuantity = attributes.getValue(XML_ATTR_RESULT_NAME_QUANTITY);
```
Comment (DanVanAtta) [inline src/main/java/tools/map/xml/creator/MapXmlHelper.java]: Looks like our handlers are mostly doing just this:
` costQuantity = attributes.getValue(XML_ATTR_RESULT_NAME_QUANTITY);`
That makes it seem like we can really simplify with the open close handler list. going that route as previously described, we might be able to move all xml name constants to the individual handlers, and would no longer need the `public static final String` values anymore at all.

File: src/main/java/tools/map/xml/creator/MapXmlHelper.java [hunk 16]
```diff
@@ -1486,4 +1381,19 @@ public static String getHtmlStringFromCanalDefinitions() {
     return sb.toString();
   }
 
+  static class Wrapper<T> {
```
Comment (DanVanAtta) [inline src/main/java/tools/map/xml/creator/MapXmlHelper.java]: This only wraps a `GameStep`, should this class really be generic? A comment indicating why it is generic and not specific to `GameStep` could be useful context for the next reader.

File: src/main/java/games/strategy/engine/data/TripleaHandler.java (hunk 10 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: TripleAHandler is a checkstyle violation, because capital A is being followed by a capital S.
Agreed, TripleAHandler is probvably not a good name, but I'd rather go for something like `TripleaSaxHandler`

File: src/main/java/games/strategy/engine/data/TripleaHandler.java (hunk 12 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: If the xml is well-formed and semantically correct nothing would change at all.
Otherwhise we would probably run into a lot of "missing attribute for name" exceptions which would make it harder to find the real issue.
So this method is basically a fail early concept

File: src/main/java/games/strategy/engine/data/TripleaHandler.java (hunk 13 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: I like the idea of having interfaces to do the individual jobs, this would be a pretty mayor change though, so not something I can do from here ^^
So this will have to wait until I'm back home unless @ssoloff wants to pick up my changes if he got too much time 😛

File: src/main/java/tools/map/xml/creator/MapXmlHelper.java (hunk 16 above)
Comment (RoiEXLab) [inline src/main/java/tools/map/xml/creator/MapXmlHelper.java]: We don't really need to have this class generic, but I like to have things general purpose.
I believe there are more cases in the code bases where such a simple wrapper is needed for enclosed lambdas so maybe create a single file for this class instead?

File: src/main/java/games/strategy/engine/data/TripleaHandler.java (hunk 10 above)
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: `TripleAHandler` should be ok.  I believe Checkstyle allows two consecutive uppercase letters (except at the end of a name) before an abbreviation violation is raised.

File: src/main/java/games/strategy/engine/data/TripleaHandler.java (hunk 12 above)
Comment (DanVanAtta) [inline src/main/java/games/strategy/engine/data/TripleaHandler.java]: Indeed, though the cost for it seems higher than the benefit. Maybe we can improve the attribute exception handling code to give a better error message. Also we may be able to look to offline XML validators that can help validate XMLs for map makers without doing it in code.

File: src/main/java/tools/map/xml/creator/MapXmlHelper.java (hunk 16 above)
Comment (DanVanAtta) [inline src/main/java/tools/map/xml/creator/MapXmlHelper.java]: Okay, enough.

Yeah, it is a balance. Generic code is more costly though (roughly 3x as a rule of thumb per some youtube videos I watched some time ago, YMMV), so generic without a reason is not good. Then again, we do use `Map` instead of `HashMap`.

**Comment by RoiEXLab** [comment]
Closing because of #2156
